include/c2pa.hpp

```diff
@@ -115,4 +140,616 @@ namespace C2pa
         c2pa_release_string(result);
         return;
     }
+
+    // IStream Class wrapper for CStream
+    class CppIStream : public CStream
+    {
+    public:
+        CStream *c_stream;
+
+        template <typename IStream>
+        explicit CppIStream(IStream& istream): CStream() {
+            static_assert(std::is_base_of<std::istream, IStream>::value, "Stream must be derived from std::istream");
+            c_stream = c2pa_create_stream(reinterpret_cast<StreamContext*>(&istream), (ReadCallback)reader, (SeekCallback) seeker, (WriteCallback)writer, (FlushCallback)flusher);
+        }
+        CppIStream(const CppIStream&) = delete;  
+        CppIStream& operator=(const CppIStream&) = delete;  
+
+        ~CppIStream() {
+            c2pa_release_stream(c_stream);
+        }
+    private:
+        static size_t reader(StreamContext *context, void *buffer, size_t size) {
+            std::istream *istream = (std::istream *)context;
+            istream->read((char *)buffer, size);
+            if (istream->fail()) {
+                if (!istream->eof()) {  // do not report eof as an error
+                    errno = EINVAL; // Invalid argument
```

Comment:
Might not be avoidable in the C flavor of this API, but it seems like you could park the error code in the `CStream` struct and thus avoid any potential thread-safety issues.

Wikipedia tells me that common practice these days is to use macros to turn `errno` from a global variable to a thread-local variable, which might mean I'm showing my age as a C/C++ programmer. 😆

Reply:
In this case we are passing an error code from C++ back into C and then to Rust so that ultimately it can come back through C to C++, so, it kinda is what it is.

Reply:
The sanity check here is: could the C flavor of the struct host an error code rather than using a global?

Reply:
What struct are you referring to? This is a function. The context is opaque to C and Rust. The API is based on the C model, hence the use of errno. The C++ API is a shim over the C API so it follows that standard.
I could redefine the API to return negative error codes for the stream function that return numeric values, but it won't work for the API functions that return pointers. I chose to follow old school C API conventions because it is the lowest common denominator.